tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,199 @@
+
+            // Technically, the estimate values can be anything, but the closer
+            // they are to the actual use-case values, the better the generated
+            // schedule will be.
+
+            // Now, let's auto-schedule the pipeline by calling auto_schedule_outputs(),
+            // which takes in a MachineParams object as an argument. The machine_params
+            // argument is optional. If none is specified, the default machine parameters
+            // for a generic CPU architecture are going to be used by the auto-scheduler.
+
+            // Let's use some arbitrary but plausible values for the machine parameters.
+            const int kParallelism = 32;
+            const int kLastLevelCacheSize = 16 * 1024 * 1024;
+            const int kBalance = 40;
+            MachineParams machine_params(kParallelism, kLastLevelCacheSize, kBalance);
+            // The arguments to MachineParams are the maximum level of parallelism
+            // available, the size of the last-level cache (in KB), and the ratio
+            // between the cost of a miss at the last level cache and the cost
+            // of arithmetic on the target architecture, in that order.
+
+            // Note that when using the auto-scheduler, no schedule should have
+            // been applied to the pipeline; otherwise, the auto-scheduler will
+            // throw an error. The current auto-scheduler cannot handle a
+            // partially-schedule pipeline.
+            //
+            // Calling auto_schedule_outputs() will apply the generated schedule
+            // automatically to members of the pipeline in addition to returning
+            // a string representation of the schedule.
+            std::string schedule = auto_schedule_outputs(machine_params);
+            std::cout << "\nSchedule:\n\n" << schedule << "\n";
+
+            // The generated schedule that is dumped to std::cout is an actual
+            // Halide C++ source, which is readily copy-pasteable back into
+            // this very same source file with few modifications. Programmers
+            // can use this as a starting schedule and iteratively improve the
+            // schedule. Note that the current auto-scheduler is only able to
+            // generate CPU schedules and only does tiling, simple vectorization
+            // and parallelization. It doesn't deal with line buffering, storage
+            // reordering, or factoring reductions.
+
+            // At the time of writing, the auto-scheduler will return the
+            // following schedule for the estimates and machine parameters
+            // declared above when run on this pipeline:
+            //
+            // Var x_vi("x_vi");
+            // Var x_vo("x_vo");
+            //
+            // Func f0 = pipeline.get_func(3);
+            // Func output1 = pipeline.get_func(4);
+            // Func output2 = pipeline.get_func(5);
+            //
+            // {
```

Comment:
I just noticed that the auto-schedule is not doing any tiling at all. It's just compute_root().parallel(outer).vectorize(inner). Can we tweak the example so that the autoscheduler decides to actually tile something? Maybe use a series of stencils instead of a conv layer.